### ISSUE TITLE
False positive for Style/SelectByRegexp

### ISSUE THREAD

**Issue by natematykiewicz** [rubocop/rubocop#10198]
I have the following code:

```ruby
self.foos.select do |foo| # C: [Correctable] Style/SelectByRegexp: Prefer grep to select with a regexp match.
  foo.match?(:bar)
end
```

Rubocop expects `self.foos.grep(:bar)`.

`self.foos` is an ActiveRecord association. `match?` is a function we wrote on a specific ActiveRecord model. It takes 1 required positional argument and 2 optional keyword arguments. Neither the receiver nor the argument to `match?` are a Regexp or String here.

--------

## Expected behavior

Rubocop would see that a Regexp with a String are not at play, and not register an offense.

## Actual behavior

Rubocop registers an offense that it claims it can autocorrect, but the suggested change doesn't actually work.

## Steps to reproduce the problem

```ruby

class Foo < ApplicationRecord
  has_many :blahs

  def things
    self.blahs.select do |blah|
      blah.match?(:bar)
    end
  end

end

class Blah < ApplicationRecord
  belongs_to :foo

  def match?(opt)
    opt == :bar
  end

end
```

## RuboCop version

```
$ rubocop -V
1.22.1 (using Parser 3.0.2.0, rubocop-ast 1.12.0, running on ruby 2.7.4 x86_64-linux)
  - rubocop-performance 1.11.5
  - rubocop-rails 2.12.4
  - rubocop-rspec 2.5.0
```

**Comment by tejasbubane** [comment]
> Is there a way for Rubocop to understand that there's no Regexps involved? It's not aware of the datatype of foo in my original example, is it?

Nope. Rubocop does not know datatypes of variables since it is a static analyzer. I am not sure if rubocop should handle this case.

**Comment by natematykiewicz** [comment]
Seems like there's not much Rubocop can do on this. So I'll just close it.
